### ISSUE TITLE
Enhance "*DEPRECATED*" feature

### ISSUE THREAD

**Issue by tysonite** [robotframework/robotframework#3546]
Now it is only possible to add a "\*DEPREACTED\*" text within a keyword documentation for generating a warning. I would like to see it enhanced in those ways:
* deprecate not only the keyword, but also keyword's argument(s)
* deprecate library as well
* a decorator to avoid modification of keyword's documentation

I am think about such usage of `deprecated` decarator for 1st point:
```
@deprecated(desc='Some description', arguments=[arg3], library_version='1.2.3')
def keyword(arg1, arg2, arg3=1):
    ...
```

- `desc`: specifies a reason for deprecating; can be some default value if omitted. 
- `arguments`: if deprecated argument is not passed to keyword in `.robot` file, so default value is used - no warning, otherwise, warning is logged. May be `arguments` can be defined as pair: `[desc, arg]`, so deprecation message associates with a particular argument.
- `library_version`: specifies version of the library where keyword has been deprecated.

Not sure how to apply the same to user keywords though.

**Comment by pekkaklarck** [comment]
I agree using documentation for deprecation isn't always that nice and I have been thinking we could add other means for that. Having a dedicated decorator with all the proposed features sounds pretty complicated, though. I'm not sure is any of the configuration options really needed or at least worth the implementation effort:

1. Having a separate description about deprecation sounds strange when you could just use keyword documentation instead. Could you explain the benefits of a separate free text entry?

2. Being able to deprecate arguments would be helpful, but arguments can be changed in so many ways (renamed, removed, added, reorganized) that I'm afraid it gets too complicated to solve this in generic manner. If you have a concrete plan how it would work, please explain it in comments.

3. When something is deprecated could be just explained in keyword documentation (or in deprecation description if it was added). Or did you have some special use case for this?

A common problem with adding any of the above information is that we needed to add them to various model objects, Libdoc spec files, output.xml, log.html and report.html meaning that implementation work would be rather big. Especially with free description and library version I find it hard to see benefits when you can just use the keyword documentation instead.

**Comment by pekkaklarck** [comment]
I've been personally thinking that we could support a special keyword tag like `robot:deprecated` to mark a keyword deprecated. It would work both with library and user keywords and would also be very easy to implement. A drawback is that you would still need to update the documentation to tell more about deprecation, but at least this would allow dropping the `*DEPRECATED*` prefix from the doc.

We could also support `deprecated=True` argument with the existing `@keyword` decorator, but I'm not sure would it have real benefits compared to using `tags=['robot:deprecated']`. Implementing it so that it just adds that `robot:deprected` tag would be trivial, though.

**Comment by aaltat** [comment]
It would be nice to have a clear way of deprecating a whole library, not just individual keywords. In that case it could be useful to have a decorator or is it possible in some way already?

**Comment by pekkaklarck** [comment]
There's no possibility to deprecate a library. That would need a separate issue.

**Comment by tysonite** [comment]
> Having a separate description about deprecation sounds strange when you could just use keyword documentation instead. Could you explain the benefits of a separate free text entry?

I like a warning message in the HTML log that pops up deprecation info. That's what I meant here. If `description` is defined, it can be used to as a source for a warning message. The only thing here is that number of such messages (even now) might be too high and the top of HTML log might be over populated with those messages.

> Being able to deprecate arguments would be helpful, but arguments can be changed in so many ways (renamed, removed, added, reorganized) that I'm afraid it gets too complicated to solve this in generic manner. If you have a concrete plan how it would work, please explain it in comments

The most generic use case for us is that sometimes we don't want to deprecate a keyword, but only some argument. For example, arguments's behaviour changes or it is replaced by other argument. While we don't want to break existing tests, we support old behaviour for a while. And unfortunately this can take a long time (e.g. half year) as corresponding teams reponsible to support tests shall have planned work to do changes.

I see only 2 options for an existing argument:
- agument's behaviour changes in some way
- argument get replaced by another argument, or is going to be removed at some point of time

For both cases, might be good to issue a warning message (again, I meant it when added a `description`) so people take this into account. Only more though here is to have a tool similar to `tidy`, but with different purpose - it can be used to analyze test scripts for deprecated keywords/arguments/libraries, and, if found, the tool reports that so we know that it is safe now to stop supporting old behaviour. As we are having more and more test scripts in our environment, the problem of analysis them for deprecated stuff becomes more feasible.

Adding, removing, renaming and so on are not backward compatible changes in terms of API, so I agree it does not make sense to deprecate such changes.

> When something is deprecated could be just explained in keyword documentation (or in deprecation description if it was added). Or did you have some special use case for this?

Agree. Read my proposal about `description` argument as a `warning`. So whatever written in `warning` pops up on top of HTML log.

> There's no possibility to deprecate a library. That would need a separate issue.

I've added it: https://github.com/robotframework/robotframework/issues/3548

**Comment by pekkaklarck** [comment]
Quick comments after looking at this again:

1. I don't like adding a new decorator too much, but would be fine with enhancing the existing `@keyword` decorator. Perhaps it could be used like `@keyword(deprecated='The message to show to user.')`.

2. Automatically deprecating arguments is really hard and I doubt that's worth the effort. Keywords can themselves log deprecation messages like `logger.warn("Argument 'x' is deprecated and will be removed.")` easily.

3. As I just commented #3548, I don't think deprecating a whole library has too much benefits. That can be discussed more in that issue, though.
